core/src/main/scala/kafka/metrics/clientmetrics/CmClientInstanceState.scala

```diff
@@ -116,6 +118,27 @@ class CmClientInstanceState private (clientInstanceId: Uuid,
     crc.getValue.toInt ^ clientInstanceId.hashCode
   }
 
+  // Returns true, if the time elapsed since last message is >= configured pushInterval, only exception to this
+  // rule is if client sends the isClientTerminating the flag, in that case broker accepts the message
+  // regardless of the push interval, however broker can accept only one request with isClientTerminating flag set.
+  // Whatever the reason if client keeps on sending the messages with isClientTerminating flag, subsequent requests
+  // are rejected if they don't fall into the current PushInterval.
+  def canAcceptPushRequest() : Boolean = {
```

Comment:
The ["Client behaviour" section of the KIP](https://cwiki.apache.org/confluence/display/KAFKA/KIP-714%3A+Client+metrics+and+observability#KIP714:Clientmetricsandobservability-Clientbehaviour) does say:

> The first metrics push should be randomized between 0.5 * PushIntervalMs and 1.5 * PushIntervalMs, this is to ensure that not all clients start pushing metrics at the same time.

So it's possible for the first push to be 50% of the push interval. From the code I can't quite tell if this same `CmClientInstanceState` is used between client invocations.

For example, say this series of events happens:

1. Broker sets the push interval to N minutes
2. A client issues a push
3. Client terminates unexpectedly (w/o sending the terminal push)
5. Client is relaunched, retrieves the subscription, and sets the next push time to M minutes (where M=N/2)
6. M minutes elapses and a push is sent

The larger the value of N, the more chance that a client can terminate, relaunch, and push before N minutes has elapsed.

I'm sure the above doesn't make any sense 😉  It's probably too contrived of a scenario anyway.

Reply:
The idea is to reuse the same clientInstanceState object between the client's request.  Broker only honors the pushInterval that is configured in the metric subscription. So in this particular case pushInterval is always set to N minutes. The timer starts from the moment the client instance object gets created in response to the GetClientMetricSubscription request.

Reply:
You're right. The next invocation will be a different UUID, so this case shouldn't occur.